AppCenter/AppCenter/Internals/DelegateForwarder/MSAppDelegateForwarder.m

```diff
@@ -0,0 +1,107 @@
+#import "MSAppDelegateForwarder.h"
+#import "MSCustomApplicationDelegate.h"
+#import "MSUtility+Application.h"
+
+static NSString *const kMSIsAppDelegateForwarderEnabledKey = @"AppCenterAppDelegateForwarderEnabled";
+
+// Original selectors with special handling.
+static NSString *const kMSOpenURLSourceApplicationAnnotation = @"application:openURL:sourceApplication:annotation:";
+static NSString *const kMSOpenURLOptions = @"application:openURL:options:";
+
+// Singleton instance.
+static MSAppDelegateForwarder *sharedInstance = nil;
+
+@implementation MSAppDelegateForwarder
+
+- (instancetype)init {
+  if ((self = [super init])) {
+#if !TARGET_OS_OSX
+    self.deprecatedSelectors = @{kMSOpenURLOptions : kMSOpenURLSourceApplicationAnnotation};
+#endif
+  }
+  return self;
+}
+
++ (instancetype)sharedInstance {
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    sharedInstance = [self new];
+  });
+  return sharedInstance;
+}
+
++ (void)resetSharedInstance {
+  sharedInstance = [self new];
+}
+
+- (Class)originalClassForSetDelegate {
+  return [MSApplication class];
+}
+
+#pragma mark - Custom Application
+
+- (void)custom_setDelegate:(id<MSApplicationDelegate>)delegate {
+
+  // Swizzle only once.
+  static dispatch_once_t delegateSwizzleOnceToken;
+  dispatch_once(&delegateSwizzleOnceToken, ^{
+    // Swizzle the delegate object before it's actually set.
+    [[MSAppDelegateForwarder sharedInstance] swizzleOriginalDelegate:delegate];
+  });
+
+  // Forward to the original `setDelegate:` implementation.
+  IMP originalImp = [MSAppDelegateForwarder sharedInstance].originalSetDelegateImp;
+  if (originalImp) {
+    ((void (*)(id, SEL, id<MSApplicationDelegate>))originalImp)(self, _cmd, delegate);
+  }
+}
+
+#pragma mark - Custom UIApplicationDelegate
+
+#if !TARGET_OS_OSX
+
+/*
+ * Those methods will never get called but their implementation will be used by swizzling. Those implementations will run within the
+ * delegate context. Meaning that `self` will point to the original app delegate and not this forwarder.
+ */
+- (BOOL)custom_application:(UIApplication *)application
```

Comment:
These are very specific  method because they are the place holder for the swizzled implementation, they'll never get actually called. I thought I'd differentiate them from classic methods but I am opened to discussion.

Reply:
Also the naming gets a special treatment it has to start with `kMSCustomSelectorPrefix` as a convention so swizzling can identify the custom version and the original version of the API. We can still update it if we need.

Reply:
No worries then, just thought I'd ask

Reply:
I think we can safely rename now but not sure if we want to do it. The original reason was that the whole method was supposed to be added to the original delegate hence breaking the conventions to avoid conflicts then we moved to another safer technic where we store the implementations in cache. Either we probably won't rename in this PR.